source/postprocess/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,816 @@
+                            string_stream_content_raw << euler_angles[write_raw_cpo[property_i].first][grain_i][0] << " "
+                                                      <<  euler_angles[write_raw_cpo[property_i].first][grain_i][1] << " "
+                                                      <<  euler_angles[write_raw_cpo[property_i].first][grain_i][2] << " ";
+                            break;
+                          default:
+                            Assert(false, ExcMessage("Internal error: raw CPO postprocess case not found."));
+                            break;
+                        }
+                    }
+                  string_stream_content_raw << std::endl;
+                }
+
+            }
+          if (write_draw_volume_weighted_cpo.size() != 0)
+            {
+              std::vector<std::vector<dealii::Tensor<2,3>>> weighted_rotation_matrices;
+              std::vector<std::vector<std::vector<double>>> weighted_euler_angles;
+
+              weighted_euler_angles.resize(n_minerals);
+              std::vector<std::vector<double>> volume_fractions_grains(n_minerals,std::vector<double>(n_grains,-1.));
+              for (unsigned int mineral_i = 0; mineral_i < n_minerals; ++mineral_i)
+                {
+                  for (unsigned int i_grain = 0; i_grain < n_grains; i_grain++)
+                    {
+                      volume_fractions_grains[mineral_i][i_grain] = cpo_particle_property.get_volume_fractions_grains(
+                                                                      cpo_data_position,
+                                                                      properties,
+                                                                      mineral_i,
+                                                                      i_grain);
+                    }
+                  weighted_rotation_matrices[mineral_i] = Utilities::rotation_matrices_random_draw_volume_weighting(volume_fractions_grains[mineral_i], rotation_matrices[mineral_i], n_grains, this->random_number_generator);
+
+                  Assert(weighted_rotation_matrices[mineral_i].size() == euler_angles[mineral_i].size(),
+                         ExcMessage("Weighted rotation matrices vector (size = " + std::to_string(weighted_rotation_matrices[mineral_i].size()) +
+                                    ") has different size from input angles (size = " + std::to_string(euler_angles[mineral_i].size()) + ")."));
+
+                  for (unsigned int i_grain = 0; i_grain < n_grains; i_grain++)
+                    {
+                      weighted_euler_angles[mineral_i][i_grain] = Utilities::zxz_euler_angles_from_rotation_matrix(
+                                                                    weighted_rotation_matrices[mineral_i][i_grain]);
+                    }
+                }
+
+              // write weighted header
+              if (wrote_weighted_header == false)
```

Comment:
why not move writing the headers outside of the loop as well (to line 207 where you write the other header). That saves you the awkward check if you have already written the header. You could even create a new function `write_headers(std::stringstream &, std::stringstream &, std::stringstream &)` to shorten this function.